### ISSUE TITLE
Consider refactoring StompWriteBarrier initialization path

### ISSUE THREAD

**Issue by swgillespie** [dotnet/coreclr#12586]
See comments https://github.com/dotnet/coreclr/pull/12563#issuecomment-312154400 and https://github.com/dotnet/coreclr/pull/12563#issuecomment-312155236 for details.

Marking as up-for-grabs since this should be an easy issue if anyone in the community wants to grab it, otherwise I'll take care of it in about a week or so.

**Comment by Rattenkrieg** [comment]
I've started looking at that, and seems issuing redundant `FlushInstructionCache` calls is a broader problem and was present before #12563

Here are traces taken during `EEStartup` phase (CORECLR_CONCURRENT_GC=1; CORECLR_SERVER_GC=0; Win10 x64):

```
 ===== GCToEEInterface::StompWriteBarrier <-- stomp_write_barrier_initialize() call from GCHeap::Initialize gc.cpp@33554
 ===== calling StompWriteBarrierResize 
 ===== calling FlushInstructionCache in UpdateEphemeralBounds 
 ===== calling FlushInstructionCache in UpdateWriteWatchAndCardTableLocations 
 ===== calling StompWriteBarrierEphemeral <-- adjust_ephemeral_limits() call from gc_heap::init_gc_heap (int  h_number) gc.cpp@10605 introduced by #12563
 ===== GCToEEInterface::StompWriteBarrier 
 ===== calling StompWriteBarrierEphemeral 
 ===== calling FlushInstructionCache in UpdateEphemeralBounds 
```

So prior `adjust_ephemeral_limits()` instruction cache was already flushed twice - first call to [UpdateWriteWatchAndCardTableLocations](https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/jitinterfaceamd64.cpp#L600) realizes that it has to switch barrier type from `WRITE_BARRIER_UNINITIALIZED` to `WRITE_BARRIER_PREGROW64` and calls [ChangeWriteBarrierTo](https://github.com/dotnet/coreclr/blob/master/src/vm/amd64/jitinterfaceamd64.cpp#L412) which calls `UpdateEphemeralBounds` and `UpdateWriteWatchAndCardTableLocations` (one level recursion here) where each of them flushing cache.

I think we may want to move `FlushInstructionCache` calls from `StompWriteBarrierResize` and `StompWriteBarrierEphemeral` and let them return bool indicating whether instruction cache should be flushed. That way after two consecutive calls to both functions we can flush cache based on returned value.
One drawback from such approach is that those functions may be entered when EE is not suspended and decide to suspend runtime on their own (and restart it before return). In such case we have to flush case before EE restarting. So maybe this part of duties also should be entrusted on caller.
If we can change interface this way I will try to hack such solution.

**Comment by swgillespie** [comment]
@Rattenkrieg That sounds reasonable to me, it's true that we are flushing the icache twice unnecessarily. This would involve adding something like `GCToOSInterface::FlushInstructionCache`, or something like that, if I'm understanding you correctly (pushing the onus of icache flushing onto the caller, i.e. the GC).

At this point in development we are free to change the interfaces (we have not yet shipped a standalone GC so our API has not yet begin versioning).

cc @jkotas for feedback on this approach too

**Comment by jkotas** [comment]
> This would involve adding something like `GCToOSInterface::FlushInstructionCache`

I do not think that this method belongs to the GC/EE or GC/OS inteface. The GC should not case whether the write barriers are implemented by code patching.

I think that the right way to solve this is to call `GCToEEInterface::StompWriteBarrier` just once during the GC initialization, only after the GC has figured out the desired write barrier geometry.

**Comment by swgillespie** [comment]
Fixed by #14105
